include/up-cpp/transport/UTransport.h

```diff
@@ -0,0 +1,94 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * 
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _UTRANSPORT_
+#define _UTRANSPORT_
+
+#include <stdint.h>
+#include <cstddef>
+#include <up-cpp/transport/UListener.h>
+#include <up-cpp/transport/datamodel/UPayload.h>
+#include <up-core-api/uri.pb.h>
+#include <up-core-api/ustatus.pb.h>
+#include <up-core-api/uattributes.pb.h>
+
+namespace uprotocol::utransport {
+
+    class UTransport {
+
+        public:
+
+            /**
+             * Transmit UPayload to the topic using the attributes defined in
+             * UTransportAttributes.
+             *
+             * @param message message to be sent
+             *
+             * @return Returns OKSTATUS if the payload has been successfully 
+             *         sent (ACK'ed), otherwise it returns FAILSTATUS with the
+             *         appropriate failure.
+             */
+            virtual uprotocol::v1::UStatus send(
+                    const uprotocol::utransport::UMessage &message) = 0;
+
+            /**
+             * Register listener to be called when UPayload is received for the
+             * specific topic.
+             *
+             * @param topic Resolved UUri for where the message arrived via the
+             *              underlying transport technology.
+             * @param listener The method to execute to process the date for
+             *                 the topic.
+             *
+             * @return Returns OKSTATUS if the listener is unregistered
+             *         correctly, otherwise it returns FAILSTATUS
+             *         with the appropriate failure.
+             */      
+            virtual uprotocol::v1::UStatus registerListener(
+                    const uprotocol::v1::UUri &uri,
+                    const uprotocol::utransport::UListener &listener) = 0;
+
+            /**
+             * Unregister a listener for a given topic. Messages arriving on
+             * this topic will no longer be processed by this listener.
+             *
+             * @param topic Resolved UUri for where the listener was registered
+             *              to receive messages from.
+             * @param listener The method to execute to process the date for
+             *                 the topic.
+             *
+             * @return Returns OKSTATUS if the listener is unregistered
+             *         correctly, otherwise it returns FAILSTATUS with the
+             *         appropriate failure.
+             */
+            virtual uprotocol::v1::UStatus unregisterListener(
+                    const uprotocol::v1::UUri &uri, 
+                    const uprotocol::utransport::UListener &listener) = 0;
```

Comment:
> I can put together an example / utility class if that sounds like something we want to try out.

Personally I think this would be fantastic to see!

If your design looks reasonable, I think this pattern could also be a better fit for the `up-rust` implementation as well. Perhaps implementing the `Drop` trait as a blanket `impl` on all types which `impl` `UListener`. Hmm good thought exercise.

Reply:
What started out as an exercise to provide the simplest possible example turned into just implementing it to meet the needs here. A couple of things to keep in mind:

1. Connections can be dropped from either end
2. A method for cleaning up discarded connections needs to be chosen by whatever interface is creating them
3. Callbacks could be actively executing when the destructor for the object holding the connection fires
4. Nothing in this implementation prevents multiple parallel calls to a callback via `Connection` - if that's needed, then that is on the user of the class to provide.

The `Connection.h` file in the examples linked below is the same for all examples. Item 1 above is addressed in the implementation - each end of the connection is capable of checking if the other end is still active. Options are provided for item 2 above (can check directly for dropped connections or provide a cleanup callable), but none is enforced.

Item 3 is the stickiest, although it is true of nearly any system with callbacks. This implementation handles the class destruction scenario in the connection handle destructor - it turns off any further callbacks, then blocks until all active callbacks have returned. This is about as safe as this can be made - so long as the connection handles are the last members in a class, the connections will be broken before any other class data is discarded. Whether that is _less concerning_ than the likely scenario that users will forget to call `unregisterListener()` is up for debate - in my opinion, the proposal is safer in most situations.

On item 4 - the Connection class is safe for concurrent access, so it's really just on whoever is using it to decide what is appropriate for their application.

----

### The examples

1. The [smallest example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMRMAPVgB3oDpkPHqSqoGBAMrIm9EAEYALKPEEAqgGdMABU4gA5AAZ9pAFYLStJg1DIApACYAQo6ekt9ZATzjKjdADCqLQAriwMCg6k/gAyeAyYAHJhAEaYxBY8qBqEPgxBoeGRpFk53uJxCcksaRny7pie5QySBEzEBAVhEfJRHpheea3tBJVJqekWGm0dXUW97jOj8eM1k/IAlO6oIcTI7PqOAMzxyKFYANSODkEMCYPi/AjXdoYAgq8fDidUWFQXABEAPIAfW0ACUAJKJAAq0IA4p9jqdzpgrkcAj5psRMKw7EdsEjvv48FQke8AG6oPDoC7oakECBMEJEC7IUgXeIEC6cDlci4ATw2VwA7C53hdJRdmaziOiARcIAAqZAbCC8wUbfHit5Sq7fUl/QGgiHQuGJRESqWkxXEYV2MWfPV66boEAoHbc/EBb36m6yWgpJjIADWFxxBF2CVpLwxvs46J9GL9FyVKe9voFid913l6bjyaVcozydd7v8tG1TqlDoVjS0op1zqlZY9LOzyeuAQDQdD4cwkeI0YuDFQBAQ8WAuZLAQurYrVatktr1eXxIY%2BDJ7xX2/e/JYTHiEHtjqX0pZqCuAFYnAgOfYrwr8QqZLRA8GQxp3bd7s1vbRxGATkJD5EDgIIfFsHLaYmBSWg8A0BAIDsG8UIBCB%2BQ1fkhQdJx%2B0HBgeVTQVtVFAEtSOJspXpQgIHZC4ADYORFCiqMlGjGXogBODkuNY1c6QZOjeMMQwOUMfizwQfgcS0RlJN1aihPopiLhYxcPhFAF9C2WgDCvEwIiMExUA4bgWD4TBBGEUgCAMYwtggJA0CwXBCBIchKBoehmDYEAfwGZonhUCRpFkdglBC9QtF0eyTHMKIrBsFBXFcBomjyPwNzmCIHCvaINzGapagUI4AA4SmyXJxBykA8sqso8iKiY6nK9LAqGJZavq/oHhaJZmrWVqKpg2Zgm6Or8tG5Yqha0qyq2DQdj2A5kQYM4Qkub0mCIFg8HsAkiROdbUUTKgQnW5pZEgo6UU2tFvTYFgSAFG7twNE77sTVAeCuytDve46Nq2jECAFHhMBBAhiEPAgNDej53l8zANB4YM0Vfd9Qw0RtyTeZHUfRukB0PWhcfeAhMAsqxKe9MGIeRi5wRhDl6cYVgrP4fgLjeYhgHhgG3jOJgNBx7s30wTAAAlrHQegNNrC4AHolZHDnCf2Ym2joPGVYuSEJHSKh0ZAT49clNMYO8ZALhRto4IQpCuf4YUAFo2RxHa0Vl/Co2lC4ZY3eglYCvqLjRvBiDN1W9QgRIgRhKCA9l%2BhORxsQ5UxyXbY3C5UH%2BTHexDDlQ%2BaC4M7ZAN0hz9AtXec3iJ%2B9IdpIC4UlQYJjwud3S7yW2KVkEIvZx1ljdoBtxE18cMfEX8%2B4QtviFQENGGjqU0ybmGiGILue4DEeEG90XUGQPAvdpQuP0VG00DuQLMFrjkfr%2B2gswjKNJ39gfQkwT5Kepr2dNwbszYEzFmFw2bI2djzPmAtCTvGxCELwFxe7iHJopSUIQcg2BQT2K%2Bz45wEDdCAc6l08jemZhAXm/NnYUUJJRASWDP6Bzlg9I4CpWyIXaA/EEv0o4YiwNrPS/kq7SxTr/DEzMOTUI0M7G6h0GFniYTggI9BrAhB4HmVspC%2BreipDSCAnCEDcPQLw6G3pUEMEgnQjSeplFAXFlYOCbCOFEPdFwnEpi%2BEWNnh1cQ8i2IXHsSg3xXgH7aEPMWdhhDiGRkst6Fh9AS4BlgvLBROoBJW32iEu%2BYT0ARMjrbGCDtEIQEcUXRwDFKlshSByVsz88jXQxKo3EDANE3QcFUzpbI1FtM0QQ3CtYTyBL1NoYgeAB6U1uIgrwMJl6MHDl4WxzZzyskvn2AhrYDwrxBB4h%2BFi8GhkgoYtxIBnoUkwHRFIGwFIrNWZeW%2Bc80GbNOdsyGez0A%2BNyX%2BAkdFDnF0WQQW5KyZSXmMUHFxMT3RvN2cYzx3onDcMSRI7AdFQnNHvP8uprzUAXLor0jRWxAXAubO/IcjZwWsPvOi8hWllnLjpbuDBecIZb1bu3Tuwpb7THQXcslhEwBgHWSGEE/AAJMHQMefgXAeCRwfseelZE8Z6j1uCAcH8cEXOIDkNB%2BcWXN23gqs8/8%2BCANBsAxmzMYR5ikRAi1HMoUcGYM4kEpIobekFa2BCuyOYggpBQmE4D9GfIJByDl/1sB5mhiEFFmTTkNPEE0gIVrIL6rZTvNUMjoHtH5sMgSLp42/Uaf9ZNgbU2yRCLQCCijEbMr1DaCAnrMBapBDiAAjjG6YPC81njuaCyub4i55mFaK8VkqSV3M5P8JkF4LgAVDDw5AKRh3/LFSfEMCrKJzvXYu65vLJ2kpRpWr00TlTzpXqYpdaoc2yK5hOu5O5a0HslKObwVABSuqoCCGVcrJVbNxZc4VNzFV6kffmqU/L%2BwaGPYqsDZ49aJHEK7fln8tU6sInqzeLcM0CRNTTCRARIEOqtTa8BRHQHzmdfQT97qMReo0D6tgfqA1BupCG7AYaO4RqjcQGNCM9TBrTdh48VDYHZtgT2ut1pp1Npbe2ztlNTGSeffcgdWMwwEJHWuiVm6RkrIbf289u6V2Do/Gu0Munt0LsvXuwZvbn1np3TZ69Ym70gZWXBqTzZX2kg/W6n9niTnEPOYB/5wGa0PsZcyzzPBxmTPYAJPWAB1EgIZ2g7FzhXIxJizFynk5HT%2BTBw4hAdjbbl0avAkDjbx5BYyJle2mRVggcyV5WMdHSmtPASvwWQCAZWqsnDtmDJGAMWZph0DJrF%2BrlMIHgs/BAy808LjaACDCASligvuLhTw7xzT/mpqXRyOr8XGs1ea/Mhgwo%2BsjquXmqLU34umyZSq1WMJD64IlmIiFadpRvlQAAdwfgtoJDYtBaq7qySmxA9rMBm0tx5firFKOwUBRFOJkV5kEaTb8ojkUBuka5/gASBJUHGf4C4aOvusMVYJsH6Qu52a86p5AuwcQSBBEukz6nR2oB0/elszb0itswB2u2PCeN8Yi/pmTQrWeMAIBzlI0rOCysCzc/dB6tP/dhnROX7Or3uYZQCASnmXsXDUAwOC66cZ04zVO37ZNhU42MRc8u8RHb7Pg6reOic%2BsAFksHcjSGpoHpPUAsGAjkS4RWrd9nHDtC4xicbIEPtZxLqsK6yaF/JsXpiBKCZ8%2B%2BmjAX5WtkB0wEVu2AjlI/B0rpDEanKZWXrGvfYk95wYK/W2Kvf324DLNz26BndMFd1wQgnumd62noUzGbdz3czeNyeP3Ishcg5IDyuhFZIDlm2icvIZwOSj1qvw2colqzYT7bruMhCIXVjyGIn9nJQNqXcr1X8r1eM5U6/7f8lDfOhHWHRSH/z1C00Lw/XEEuX5yNxNyizjWIRbj2gOgCHDVTVt2F1Fy7TzzFDHi0BXCl0lFbEQP2m9DL1xEr3MT21MyOQUUjRHQVg6wySRnVjRk1ixzoHQTwzNUI3tVAVtXI05gX1gQRmFlFg%2B3oCp1Ti/ylGCR/DzEsXxxgRoS5mJ2R0/lbw02iR/Bx2oIPwIJB3UIJX6S0Nnm/CMIVjPGy08Vyy%2BSsQJDZBpXEBBEVSsJ20oOr323sPoP0PqSLUTRLRaXUR4AOyMOcJrU%2BC61Kye2ZUcUlmRU2xAA%2BRsOaVngO1nmxWISSKrw0IO1qUdQTWYACPMMOk6WqRZxLlMJAGOwa3EBmXO1azriZ2u0cIYBBDRTuCJRuyvXvFCISJC110/3a2N2eylAAD9YjJCoCNcAD/lRVf9dND9%2BsLgAAxEgfYHGLBYHbXQgIJCQDg/vYbCZDGf5YfV3flCfO5F/FokEJvTXa412SCK/aApVJ/JYoEPwwo1%2BDkegbkJaNgSAi4EMUcf7XfLMBIIHNZXpIJYI14l/UI24ydZUM4VpDRG49op5Vo54x9UDOAs8LDbeNuLja/Wor0U8JnSDBHPqUVLBSGNAC6eSdESNeQfUevRUFUa4idMDRgpELSJYgmVgtEdgysJlRWPWfkomJ3HSSwfSQyOKUgUySxYKOyYyRyJAaGSWTyOiN5GRAwV2AEAIP3N4AAaWwBBCcDUEhBiABBBBhAAE1tB6EAQARMAUgQhgATA0ALI6B0gYhrBgAh5gBMBIQ3R7BnA0oWdphw8gQWQusCBljvTkYDATBvIUZ4AIAH53JiBgyLAUzsz6gUzEyuBeABAhARBQB9ATAEIDZgYH45SEI/dDwWhlp9g6yNB3jGA5SkpgBsyKzHIMzt5syohcy3QhyEyOZ/IWjgpyzKyNBqzUQ3RjJyANAGz4hJBmyDh5AZz2yjJjBLA/SeydItgEIDSK8dAl5TBAoDBNyzALAEh/t4z6BsyjhdIZTSAdyTIDAYQcQHpvhrzlSHItgQwQArwxI9J9BFATAWAFBRI3y5TTIKz3AQAxJ/ypS4BYAYBEAPQvT6AyAKA0VsLJhgB5AjhRyq10gvwIAUg5SUh4h2gBQryTAaLmBiABQgQUgLyvAGLSBPS2AJAgRO96LFysBXTgBHEvx9BrysADwbA5AhK5VBgLlxLdyuABgWQDhdyuRGhOy8AUgYYWKggsA5ToY8AoKJKTAtV24tBnTpLgB4JkoVTRArB%2BYAA1PATAf7bcrisQUKV8CKZQby6KHQPQRchKPc5KUMlwZwSwHSr8SALYAo8S12V2V0Z8CK1wBwQwbuJ0k0KEWEBEEyLVcZLAWK48dqPqLKQIcaeYAqdAQaEqeoUoaqfIKqnoBqJquq9YMq5oYYMaQoVq3qbqgaFYYqTq6aWqeoaaDquoLYdUiKF8/QAyWCxc%2BCnkMqBiV2BiRQC4YAZAG2Yi/gBwRUL87OY4TYEwFC1UrC2VHCzUz066wi4i0iqHCiqixcpiuiri96litiji5U68ni%2BXfi1%2BOU4St0sSriqSv02S3c/AHEBS1MxclSlnSmOsw2MC3c%2BCXSuigyjc868ZUy68iy7ITAayqGycOKLYKgJyjQVy9yzysyqKMKOQSKAKzQIKuU8weoLslKMMqKzGkq%2BKj4xK5KohVK1KZwDK7uIEDQAq9IIq1MuKrqzKdM7KFqiwWIYauaBqqqZocatq5oKaqYRoRHHqzoNWia42vqU2w2iarqc2xYEYG2ma78hQeaxa98%2BUgwVa9aza7a3ai4faw6iAY6n8o4M62yCmwC4C0CgwCC0gKCkCpa3c%2BCkwL8eoC60gNCzCz0WM268Pe64gRIcc/QTgNajaranavao4A65M1QJmvyqKNm2KEKuqWyb8mkCwf7GGHgBit22U5agwaMggWMvOf4Uun2iu/2wOxUIIAi4sb4BwDYc6imzOpAHOlkPOueouvyEusu32yugO6uhwWuny8KBQfy1QJu4K3czmtuyWDu%2BoLupgHusyvupOj8/QIekevVce8uv2qug6me/O70%2BesOpeiOlUrYQ%2BCVSYUqsCuOhOsSD2lOxC5CyO0gICxOsCo4fu5OpMiBgC6U/QY%2B9%2Bz2hCi6rYNDTKRQIAA%3D) shows the creation of a single connection with a small lambda as the callback function. These callbacks return a value, but returning values is not required. The callback is invoked via the Connection object a few times, then the connection is broken by resetting the handle before invoking the callback one last time.
2. The [next example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMRMAPVgB3oDpkPHqSqoGBAMrIm9EAEYALKPEEAqgGdMABU4gA5AAZ9pAFYLStJg1DIApACYAQo6ekt9ZATzjKjdADCqLQAriwMIA6k/gAyeAyYAHJhAEaYxCAA7Mo8qBqEPgxBoeGRpLn53uJxCcksaRnZ7pieVQySBEzEBMVhEVEemF6FHV0ENUmp6VnKGp3dvaUD8%2BPxk/XT2QCU7qghxMjs%2Bo4AzPHIoVgA1HYnAV3ETACet9h2hgCCp%2BeXmDcODiCDASw3E/AQjgc7y%2Bn2%2BVCwVCuABEAPIAfW0ACUAJKJAAquIA4tDvgwLiFrrcAj45sRMKxXiSHCd/HgqCTPrSQl4rto8gQUVQaIcbpkXJ8rpKriF8jYrkCQW0bickVcZLRaCkmMgANYaEAgBVDNpU2jiYBXeIEUiWiQ2q2Mk7imEfKXy8SKwoGgAS1nQ9CuGhCKQ0yGIeDSEHyAC9MGiCFcWEw6ClUJwbUbQRFDbJNdqdY4AGxFtUpLai51ut1sq4QMBgJMptOYDRouwAVicjc1aY7SPLdjF0KrI6uTBCRDV9GsIR4ytVHa7yZ76auBAQeA0fYg46IW0HTkTy9TnBbbc73ZPff4dK0BAg%2B6doqRt0ro7du9QN07CBt9nbC4qu6wLGl6HBzEwKS0JuCBRgQ6AGiwqAAG6YBAyBljacwISASGoeh04MLOWyPm%2B76SpezatoulGcH285qq%2Bw7kZKdIEPsDBXBCTrMSOg4vhKo5sRxFb8UxgmSmJsISYG8EGqgPBtLIprmraBCvGqJC3rkDDoFGeCxvGR5NquVpXKZEhXE8A5DjJ1aIhAtFnjRx69gBNlkeRwnEJxEAAFROdRF6uXR7kQKu1nia6o5SdFI7eZxB5iTxMmxdCPDhshTAEOwvGyThDzPFSmZtAa9wapB9A2icGmBW2KVfJkL4NdCXI8piLY6VgxCiTJHUaF16QQHycyCsKmAlrkWEGXGCa0fudlSiAxkrmijkhTseUjgga25PwQYhmGEZobRNqLuum7bmZFkJtZFZXAlVziAAsi2GhMMAaERaRz4kaJTUcnFVxmnKZkvW9H1odd9qWXdSWLZKNbrSZaLzrcqqGB5W1CZg7E%2BbWnAALTyOWfm1i8zgk1F778VcLRaJaDm0aj6Po1cJO9UD76PXRDU001dO0Az8NczjeOcS8fMxQDqUyy6GV4FlOUgHlaAMHMgYzUZzPU1KI0CkKeCHGVHqgb4IC%2BrpAY7VFyXOtCyGoHg6BrrIOo7hOqCTagNqfomKS%2B575mB5OcO2UDft0nwTwMXtaDENp4h6fUNrfbrknMjWqIYji%2BJEnl2EGmgE7KgEVL/A4HwaDq8QWrRFel%2BX9SN3cDcAO4INla5fugX5t4QCAN1S5ecC3AQN36Q93OXMfD63hccFb6eM7WUe0GHnlugvxcJnP4%2BQm6hIoniy0H3vVx%2BWvs/T/PcmL/6y%2B0/Tfwi%2B%2B297LvN/7wCbqJCiVyYmwNoGIABNWsLA8DAAQAmNIdNUL7h/ufBe/haCPzlm6U4rJ2SwjltCMyjYGAPk5m6fWY0jZ/FyMvfqg0eqyHIRAKa7NSJ5WoUndIVxUwpAYT7K4DhmEyT9pgVCDEF5Jh1HGDQnc6ToCpKw3S6RXjcJtJjZenRaDu0YYYaqNoqZS0lGojRPD5DaKYaot2SjeEmN0Z5IRmAbwtlxg%2BMx6iLFaKuCcHR/DGpIn0DsWgBh2wmAiEYEwqAODcBYHwOxQgRAEAMMYHYEAkBoCwLgQgJByCUBoPQZgbBDSmyzOCFQEhpCyHYEoYp6gtC6HiSYcwUQrA2BQK4VwzRWheggP4RY/R2zRF0hMOoDQFAnAABzlDyAUcQ3TIi9IqJMhgAypgZHkKMtpZt2grGmQ4XpgwsyjG6IsjYyzVkQQWMEPoMz3ArEOUMlZIydgaD2AcI4pJySUjuNlVAED7AnDeLCTOZJfilyoCEMkSlUG/KZGcQFFIX53DYEhYgks/lfABW8uFQRFKFGUpC/50L0WlwIE8Hgs1HiEC3Lil0uTOraj%2BOqPMuoNCc2hNSgatKrhYE6HQTmOVIlWBylSIlJLqUALxDaIVjBWB2P4PwK4HxiDAApSii4TANBMvKvQTAlt/QTRagLAA9Pqq4rKeDss5cuQGhqrjYgkOkKgtKVafCtZKMmEFvDIDphBKCMEIAyv4OWQmao6TZT%2BJPR6TArjavoPqkqhQrimrwMQaEzqpQQD/nibAy0o1/E3FcMQPV6WYD%2BP4J6iJ6Val1BmApSp81qlzOw/wC0PgprJgpdInyeqpmCEQwNsbxBwNkCEENTLJz2qFn8cQIp1x0urXG3NKRiCoHEQwZNRqXVPRJY8IgxAe3ylzCOhAoa1WoGQHgENLty35lrDWNWipMDoB2Bu8F68Hq4w4rXMcVwsqhAmp8XlfAQ2CuJZKtgorxXAepX6uVCqlWtQIMQbkCY%2B2JXDm6GUH6NUVp1CIu%2BIKwWFFkXiCA8rFV%2BsfG8PR0pZQWmzThnCkiuj3rRIpJNdxzV0DKvWrVfp6CEZtCRjQfrHQUc8uhuUARCKzjowaPDWYqSO2dnBejUimMseKrO8QrxyPLzExaDVlU4WqgXgx6RzH4PqZAnJyFlHdPATvegbQyYepswXuxKJVJs0ZlzAZx09sZJuqNnZ4097HOJs9Z0b1kiICYfzEWEsGEsJ3wUuC4qkmeCMgcMWTLU56RETnGzJKTUsYI15JlENQI2oEDxEuxg8avDLw/EHS9uppO4SYOItEJn73FVzFhxRoiUJoQwiRBrUo/a3vWa1sREiVMyLuMhxRzWdQ2h4F4Lx74/adytoZ/KiF2szcY3NgIThGPZsWxphgf5ev5kSzhPCQ20uPtWwQdbYsRIHi2zqv8F27Z5TSjJNtW6SAcNQN28sasNavze/jesS20T8DNEwPSWx%2BBcB4Im%2B9TjKP/aBlajq4sP2oWIPkftqBESA47U4mS/7%2BW6oCBKkVmI8R4gYkz8DwqpW7cXgZtEbJ4xUnrAvTcnWpVomQoRsVX6nZzewDaLtELsAMXgyEXVKKt5JaxeIHFAQmd4g0hT7dD5iMwag10RVxXRaSgXsl7FEKdfM40reEItB1ItRK0jQXtjiBojpAARxV3MJjFuWKSnG9dlrbM4cI9QEjrHm9RxIz9maXUTGMIMSW9H3UcfgYnvEegNEw3iEh7dE7l3DF/LJ7zwXssO4TcyteyxHHxeGCoG8FQJ4vOqBojRxjvSA38JLZG5RjBuCSuPVL67zyTepRWsSOIQmCVCfpBJ5xMnG722G6bW6GngG7gM857r1nkv9%2BgeQcwKCcY%2Beu4CEL1s702Bi4lzaBTMu5eg4V0rhDqu8ov/X0DndWwxupGMqY4MGweCeDknuROPumA/uLYOU%2Be4BLEYeGoWG6e4eOo8OiOyOo2I4ieQcleqeKQ6BqB%2BYme7sP0hB%2BeheUOxekoFeueRBgBZugm9euBVY0%2BLELebeHefOPe0iSmiEg26EGBQ%2B8ekk6CUoOOCsSsuUMkVqAA6iQDqF0HsLpHmsDsZrNmZj1H7iEImh%2BhGjwMGNBB6hDsrl4CQAXBYQmNoGVjlBVjYdVsuoVs1M6MYd6sgMtFak4CXNqOxLmDHHMHQLQPGvYX8OuNYHqN3GuIeryAEHrjJMhoISAF1vnmpvNhgRpAlqVorOVuIJVs4YwOWMtHDuhGWP9AJB8DISGo6i6G6FaniHERqkWrRrmrmKgG3PejETKH8FoETkQpODlMQBAswDlLETOpZiaDJLZidnSLRmzOxv4jmBqK0TxnTmztBsAfwL5nlFQOGCWnMdxttsvL/v0UNEgVWONvsHSBINXiQQypgdHrHg3lWOcd7nofAUxp/iruwR7mAMgDcYwAQNXqjpwOjgIX9LQeRFHm3MmPeICQnMCdXq8SPlUWiYDA0UamoAwFBLnkyu8UQjWLmHWqQYylxEwMIjQAwDBN1vIUaumpmlcM9DKDAnSvWi7PsV8raPkNcBGniS1pEQmJ3EysgIeinjIvSRoT1FAekDAXAYHvnj/tLsaq3myLwV3vwZjgvF0e1joT1mSQWJCplvFhUdCViXuoaRSUyuIC%2BlqS7MSRqLEcGugEyp3MIlwIQHSZblcFatOmFvShwpXrKh8AmEKfGk7HaFcF0XWpxLeLjBMdGfSAWCVlarkFaOwo8rEV3ISeDtYNKLicGdjIjA5BhGCRCZjlCahiHmWfGfeKiaOHDunikOwW6FHtweqWiOIGhA2RIeiVIaPkDAvJ8t8lSPLhpO8fKQHggW2GKGOloL9v5nfCOUbFSDqcmfqZkYacJqSY8fVPbADG7h8CamarjMuDypgHyrvvThBgfsfreWwFBgJoyJ8CqmqpaZqgsdWZKLZgqAxAtncJsQJkJtZqJtRh%2BWgWzAqJxtuTZuBRJrllJlBR6GVGlrbEucpodpuQEAqNkRdvudYZhaZhkThVkb8ruVhgRRhfJJrswHbghTOOluRRcIhTwFRfLCYUbHUW6C0ccTqikWkdhbhcxR6LdgaIJSRTFrqNkQHFzjblrvRWhcaVloWGqCEFWsCAaHYXkQ4QUU4TVgwFvqOKUfhehB6I%2BmUcNn%2BGlmtP3kNiEFWS4IOW6AAH68XZpELmlShR51lx7Fm%2BlGoABiJAhwTKMoMRcJhABZ3goRJJ/his7JhpbplJfwCU3p74SME2WYaIlx74WVbQaIhMrwuZ7BsU74VqKItFgRNo9ACYjybA3ZVwOoLebcExMcCQ3Rk4LF%2BZs4/lmVNluVo4/k3VeWOVZlUxhQOVpVkhfZmJUoBuwO8uRC5hRe8Ub6%2BM%2BVk1/AvRBeah9Zyoiu8g/wKltYfkm14gU12OuCh5B5qoVqJ5IoSxgMtM91UqbKIoS2GgvilgASQStSpAYSyGRScSISiSSA8GRamS6E02AmBghMSIAQz0HwAA0tgGiE4GoNiDEEiGiHiCAtoBRkiEiJgCkCEMACYGgJEnQOkDENYMAEOp9NiAhPYM4K0oCXMF8iiBOMYQQIFdTdSgYCYNki2PAJ0vgNukzRYMLZLcYsLQLVwLwAIDEqQKAPoCYJuDauighCEuQBoM9MmO0E8sbDrZuJVYwP9Y0sAJLWrYkveuksQJLVENLQhE7fzVKvkhNWCL%2BKrerRoJrb8NrcYLrfrfEJIEbUcPIL7WbcEkHZbdbb4jsJuIjftnYagKYMaAYJHWYBYAkG3HzfQJLScH4r9aQDHaEgYHiHSC/MyFnSDQkjsDqCAO2Fov4voIoCYCwAoIYFomXQDYLe4CAFonXd9XALADAIgCgF8ujvQGQBQGZVTTPSAMACsq7S7svpQCkP9SkPEF0E8JnSYNvcwEiiiCkOnV4PvaQJTWwBICiAwOvP9VgKTcABqvqPoFnVgEmDYHIDrfgHSMMKhK/UHVwEMBOEcEHRma3bHRGI8EikEFgP9fBngJ3W/SYETqmFoMTZ/cANBE0qDaIFYIqgAGp4CYBtzR0X1iAlLqjlLKCUNVI6B6A631KWB03NKs3OCWARj6iQA7DyXqxw2EzYTows0uDOAOCGBXDw3ZxYi4gEiJCEihJE7hhYDcMPhrJZh%2BC6TTLGKxBrCDLTDGJzJtDaPjKVCFA3IGPqNtD7I9DnKlDGK7LWPXJ6NLIWCnK2MlARAOPOO1CuMkykAQ3lLF36CBKl3/VhL6DmQjKFiEyFiKBXDADIAeorL8AOC1iV1Fr/AnAkwmDD1g2T0L3pBQ2U3T3TDL0nCr3DH6gQCb062H270X31PH2n0Z0oOX1fLX0Ch3170/0k1k0v0X0f103f1B2/2gQAP/XAOAnKwm22qQMmDQQLq71wNgO5PhjINZ1oN5CYCYPDO1y1I7BUAEMaDEOkPkNtN0OlJyAVJ0OaAMP/XmDGKW1sOiNuCLOqO8O0Wv2EyCPwTCMtJiMSOEwojk0A1KPOwi08NWMdJdJ2NeN9LoAWPLKmPzImNGPmMuNHJuMtDrI2MmOOMjA%2BPrC3JXJjD4tEv6PLI7CBMKDBOhO90RNRMxNxMJNJPswnCpPpNV1ZM5MBMHMN1N0t0GDt2kCd3N1hM60RMmD6jGJ5OkCj0T0fw83FNT3U3ECJDu36CcDROxPxOJPJMcsOBC2qBXM0OVJ3M1JMNlCBPOwWBtyPA8D710t/WSsGBc0EA82lpMu6ussGucvRaqsz1ZN8K5MHPytIBKsTgquFPquavavMt6tsspNGuVKmsKC0OqAWuMNB2PMBNV22vGL2tMCOsoPOsStB0RPuuetr7xs%2Bv6vsv%2BtBAxs8tbChug07CHpI7TBqOt0itis93hP936hD38ukCN3iut0nAusVv915PBMpsMuzujtE4r4gCKBAA%3D%3D%3D) adds some more complexity. The `PostOffice` object provides a way to register for and deliver events. The `Responder` object registers for these events using one of its methods as the callback. This method accesses members of the individual `Responder` instance when invoked. A few `Responder`s are created and sent messages, then one `Responder` is destroyed and another message sent to it. Nothing gets returned because the connection is no longer active.
3. The [third example](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMRMAPVgB3oDpkPHqSqoGBAMrIm9EAEYALKPEEAqgGdMABU4gA5AAZ9pAFYLStJg1DIApACYAQo6ekt9ZATzjKjdADCqLQAriwMCg6k/gAyeAyYAHJhAEaYxCAA7KQ8qBqEPgxBoeGROXkF4nEJySxpGdkemF6FkgRMxATFYRHyUU0t4m0dBNVJqelZ7u2d3aV90yNjtfVZAJTuqCHEyOz6jgDM8cihWADUdgcBLEw8l9h2hgCCh8enmBdXLCEEXPePLwcRwYJxC50uAQ0CA6mHQAH1vr9OP9nq8Qe9PpDMAQUYDgaDwVcNARiPFgLiAWiCR9HA4ggwEoMGPwELTKaigXgqFgqGcACIAeTh2gASgBJRIAFQlAHFKZz0WCaVcfMTiJhWBSgf4ueyngB6Q1Gs5PBhnJheEKyM5cXj0M5Go0A34sPhMX4QggATx4jFYmH4gZNxGAGlxJyYGg0Z0k6QAbnhdhdMi5nmd02cQvkbGd6YzvOJPnyzjJaLQUhaANYaEAgPPNAsMCFx1B4dCkYOhwP8f4HVMvNMZxFMFL24noWtQmHwxFcM43Hhw2ioZCVuGXfsZs7j2sLiE7kBqskdg9aHFEggTw/Q9XwngkiH1pm9h4HbBnWh4YmMdIadd9vUtxbNsznEPkvzQBkGwgU8STJEtxGJRwADYziIHhExPS9JxvWE4XvYhH3EfNCnuBCGTWZNNy3DNXm5TBeUFYVxSlWUARojMDzQH5MQhC4HAcMCIOI6C7AAVgCNkBN4q40NQDDkBkgJ%2BIccTJIEjtaSU8izQhPjaTWLS9Nkg9/FoDd2I4/iDh1KhLI4g8QgYPAAEcQkwJcV0rD8IAXTzVzhQyAMHDiuTOCAmB%2BVAzgYDBlWLT9vwSYg/34GgGHQCB0MTIKnBiuKzjAMBLgSr9fmS1L/AgNZKLsFN7Ks9NIqIFCdIbQpoxK/KsAAWnuLRIPQCyQsatqmQ0fh0kjTAIEghhcsA0b0zC2aRPGybXW9aravqkalvTRLyt/OFJuIaasvknLhqefbkz5Bqtzq%2B6RqevUeBCUdExABqs3ggJZHLKsi1zNbG1rf6ywrVdrq3X6cwACWsdB7S6p8wZARGMvoa6Gsx5GPnVYAyvSOIkvSGDsMPOCc0gpCHFQ7LkA7CHAeh%2BnWuQFIduojjmui8S8uhLHMA7Ob2sLcTi1R0HClrTBiRHRKEAgB6rIPFhUDjGbOY2C4xKcAgEC/DtGclimrynW98IfK40dIt8dO51WlqEjQxa8C6FNF4iFr2qynt9gcbrVymnNc9z/O82hfNuSPAphqzDp/FL13102xL5fh4i0TpVqgplA8a9UCG2M0hfxhP01e1ERuA9BQIYbAtYkc3J2p4ByLphnLqZzsJqDDpQ0o2mCCo53t0py28OXVcfL8me10LqyVr5uSFKXYmKuBpOKpO9LMsZpfbvTRmN7JlLCuKg5SvPyqMu2sea%2BDpaUlQYIzjEJz4QVlHr4/2QtCV1GnzVqAs16YQ/JvX8espZ/wAFSnx3r%2BIBjUxDEHCiA%2BmY1GxnBAJA2%2BTs/ZLQgHA92jY1gRRDP3fgR9RrV2fjRehj1MjPReCwnGzxHTGlNPXJgZwUhOWQAgUCvJiQhCoLyIgNpODpGQF%2BD4hAHRcP1PKfEGJPQIHVEwIab5VFvCVJiIRxBxCoFxHDDuzA2AaB4BaD4XFNEmI3r8M6tAwzBSDrGYgCZdgQlfsEDs8QcTswdm4/sAI65nHQKgJujACCtypqSGmiEgmoRFvwt%2BtB%2BEBIkK1PAhCGETXEDEluaSUgBMDvQgEq9CbE2IKTI6xB4lHiSQwLu4DkD5K3MXUu25%2BA1O/HUqBjTGaaX1mbPxmSylnECbkzpy9eQQC5vxFwzg8BAIDjjdhT8IxRjODEcQwARROWcjYFmUNvJ1R5mcCJy4bBHIZGSRZGSsnTJyVgvJj8GGPU5PRRiQpRQSmlIkOURCJ5Xm4qPYyylaT7JsD1Ygxz4JGSuHxFI2kjLSShdM7SpksYoM4pTCF6KBIxBFDGSUTxJRqEkHgzm/AsweW4i3Wq18urIskrJOlDK4RMriSyjl6kogqXxemLiWxIUotkrSOM0zOqYslcpCAcZ6VaB5VsZlnx3yGE%2BAAMRUsSOgmTgBvyGtJPBtJ4hxlkG2AyOLKZmRFWC2sRKsUwrJT1fKvw0LQlHjIZgxAvSypUtpWa1gOhehVYy9VfLNVnG1ZcPVtJrVay0uagSWBtG2qxbi5GjrDi2XHitJZPUeprPcY1A8hsvxwkNloi29BMCLjQRABwYlDAsA0LQ75Rxfn8n%2BSxIFIKvkEvBeK4lDhSXkspdS2lKRI1qqcjGkqbL5UCpLHO7lvLqpKVpJpASjqxU8VdQJGVX5g1YqVfOrdLKtW6v1d4MsZxjUYFTSpS11rTUOH5XxHN5ly0h1HUehVKlJ0etil6w27oSxhoDUG9lfFQ3%2BojZu6N26yLxoOImgSybMCvtpBmz936TL2rxf%2BhyhKx3HonWSwUiQHirp/SR3NZHGHagyrqF6Wyg5blPLhO8NsAgzIdnGIBhoQb5xweXeg0YoRbFoLw2gAB3JgXpoxpG9R8KwxJ5yYDqOka5HQ8CK3lq8s4XotjoJ2WGEaWATjel9DBNYtY8b2k5qJ/UJoGBeirTmWzVhbzmioM4jTLB5wqf4R8AjZxFMIEYBp7BhQg1aC1reBqvHpzW0IiqHJDs/Xho4U8d6n1kDfRGrCw5iLTkA3OdVceeC4zxJuJWDyU8dGCZy9gCAhgaqkHHumTmME%2BmYCJgM%2BpycVYCVoIRDSesDZGzDBnJ578pkzPeYZFMtyKsPJsIs8pFkWE9b69BpDjWmDNbhK1iEQnOvdadlxvRvJ5ANQAH7lfuSc4AZyqxod2sOquPyeR9uYoCtioLD0SrXbSV7Bz3tki%2B6uWrDHiNXgdSx7tBbOOsPzexuyqJ2HuPCa2eusnFNwi4L6D2gTpEiFeb67YxA5n/Z7YDpiALWLArSxRoDkOBLYE4BTyFSPlLk/HWmhw2lkD07tSj0jVzsf4Fx2wrHzwqc3HiD9q5q8ABe6R%2BZ/36c4sbyUJsOF%2BHTL9QDV7iHimcA3JMhkm7N0Ei3LG7eDPPib8DCJcMCQqSNKneBgYHCASTsn/PoIHD2wTmz0Tm5xNpE73d/9XFpLySHhAqBSfk4j1HsJMfinx4EonmbJJ3K57SxnrP4ePZiXL0/LcOvjFDbPLVljUSC%2BO/ls7k2CLU%2B0ND9nj2yE6/cYzO3uPnvUAEG90n0vff0%2BZ7DwLiAw/pl%2B4YWJgIsXZ4pC0ZWeChsPhkMSzFugHw%2BGllZt5M9CLtvkhGqvKbwN1ZnZa3xiEb3Kufeq1We4rermVqaIaj1zx5jL0xrKLbrZODj6xIm5Taz697l4HYh6UxVp/i1rAGTgNpNokAQDyCGDtqdrrIA4MRA5s6Dqc6AYQ76QCQiiYBuRd7biYApbwbI5yyy4NTy4cYMJTbN7Yj/6cEkF/LA7s5Do8Zc7UFSoCSxgpYISuj0AehC5OocAcEvRsYK4NQED8CmCtgMACEjQD7V5xK15r4FZbib7b7eS74aj745iH4JaFhUDGKhYahCI2gZScG/aNSrwCwICiySzAyX7nI1h1gyziDNhE73BywKyfRQgqxjKQGXK07TIbh3Rdq0RCFkEDqg5/bKEurAa0guYKJyri4XoQDCIoStRwIIAspYYODPqfq4IqQViEbjrmgZTiYkSFhnqsGKqzT8TIRVEdJ3q0gNF4YCQtFZrAa/oHoSHjp8iYAKEfBSY%2B6lHTFMZ/pXLo447jwIB8F8p5qZGs7ZEc5g5zFUZFFwZKGIADFVE1EjECRjFmrNGZq%2B5tFIydHixmg9HXH9GVFYKkK1EqRPHi5i6TFvHZobGHE2Q7GY6eFbEZg%2BH6x%2BElgBFdRBFVghF2zhFXB1xREcAxFKzxEuCJEpjJFlp5QbJo4ZHM6kHHEg6nG5Hg7jp0FnhoRd69HKGo4ImirnEFECSXE/FrFro3H/GoTVFAmjEmrjEODglfrvEdHYnfElEhp/HswAnDEJrAnSnPG0hylEbKQzHUm8lUHzGLHYhH5hFNhKFGk8nWQY65HID7H6G5F0Qs79oMliEAbOqUb8kOCCkqllEVHqnin3FalSkvq6kTGvHylUbtH1xKlXHCkIZqmDEamSmPE6mgkvGtGQky7MZ2lcGK7MLK4b4eZb7NDeTdLEAfbQaQy/6cZ2lIlOAon2AZyBE/6rhYlWlXYdbRHtCxHKwCxmxQG27Yg9I9QACc05qRVJTZUUtuwMJCHSLpjUbpdJHpohhaCyDOnyt0zJVG8OVZ45NZsI56wGcC6CeZ7BBZ48T0NoKee5%2B0B5fpR5Y5Jcp59c4GRsNgnJtpd5XGa56h3BJZDUFhlZ75pc8EGJs8Hqu%2BqAzWZox%2B4RjZ48zZrZaJf8MF1Y4MPZ2W542A/ZisX4Q5CRfItWZJ1ZZoU5M5fYaRjqex6oLe6RTUC5V5f8y5FCLF1kXI7pIhFBoKK0u5lyR2JpPp3ONBdInZx5H5CQ9cnJl50uN5mxAFxYixWgT5S0L5POUl9Zs8VFZ535ZIf5UJxpd0d5wFxZtE92LC%2BgGwtABgYkJgEQRgJgqAHA3A8hAYQgIgBABgxgGw5RKAcUuAhAJA5AlANA9Ali7ASpLIKgEg0gsg7ASgCV6gWgug/lJg5gUQVgNgKArgrg7g5pTIfgGUcwEQra0QGUywEwGQ8gBwAAHOUPkI2BVSAFVbkK1YULVXpvVU1cVZ4I2MMLMMED0B1bXgMMNTMKMPEOMH1QoANQrKNSUJVZNTNb1asA1Y1RsBoBZrsAYFSOolcO6KgCwImFqGogYhCFQIIo2LIJdfooSNcLpiQF6I9YqM9fJPdX%2Bg8ByFdc9fZh5CSEwIQG4n9UHDFdYrYnWVfp1L9gCFDTYkmFgO0HQJ8i6G6B6FcEDTFWcCKJKCbD6H6GwN2H3OGFptGCzJgJgEUZssWGJkjTDajaDeZE/GJmKBIOkFQLYqVgaB5umHAhPO6ImDaISSRRAN2JRB6sgFol6h8VReaGcEUfqImTYngIRJwgLRmBAIkAKJKIRcrUjPaGemgrDTTe4fXKgLyNhczFaR/CQObegv4IZFrRmELfJFNEQOghMtuh6omcwbIFaGbnJMnhpeIEmPYchcqfwsYohQCGJoLaBL6GdN7X7bmADNGPYZGHtXIu6GedheFGFGQrCLrN9YUADIGlRfBHwlaqELhs8JjVYNjQELjf6PjYTWhMTTFWTU8FQriGqCEF4J8UyJpZmNmB3G%2BV1AeLdSCI2BCATRAH3V2IGEFK%2BFcuYkbcLM/pPHxplhCCzXQODADDTXTVcATR2MvdQi%2BEApvQEPQNYCEDwDvVeLPc%2BLiUTk0nvQRERBJvbNgGvbfRPRnWWMZi/ThBlj/bbHhevT9MA2jLCNoKDexcWJWk/djFcEUczADMZr2HngUu0N4IpAg%2BgEgxrWLQOUSUeWKeulhFeOXeIA9bbA/QwE/f8CGSWCw0/cDCJQdmPVuNoKSFar8PSIPV4JKAhXFjwF4I6qvIXdPZTE1m/tOI%2BNJX/urJrNrFzNxXI/bQo1eEoxdu/tA3/TiZ1rbWcNIwQDowuSseAyAIY5dlcE4DCGfeY1aaLNJXQ7uJo7NFwzwLrFY9xYrZcisd7KY02PjnLvdiNJ7anY7b7cPMkvwxmIrUVNhSdMuNotVJNPzhrbCKudZaWeYR5nQbJfBClvkIWNbcnV7bga7Qwk3fnZ6N3e3QTZKMDBfV3b6HjbisZnCFyDWhCEVAeNWhoP6HCCJufZKJ3XiW%2BB2BMmRF1HPriOIfQ/eBXX%2BgEO02RHE6dY0hQlfWTYPJ2ik2JSAAw8wFszsw7ExSELQOePg41CtCM8wekHCOqAwd%2BPCIzsAgufI1hdJZk6gNk9xVuCvAuQvHhJzB2XpZWPwAvP/h%2BF5NC0siJaCkXPLPcxKsWCQlC/CDrJQivTQo6qBfXvtOBlyF6AM1QEvvk5lBo1rKGnCzVHmjE7kYrXcw8%2BsuyyU2cIkOIPCiebWZU4ljU3s2nQ01uE0y3W3WwB3R011F03K3YiRv04M%2BeJCJTGMxM1M9szMx2HM9gAsxkks3/Cs7orXETrU/Ewc0vVQsc1Qr8%2BCwsq8ylh8/Qe5N84FGczRHI9JbC1fsC6C6SxmBC1IvizymiuidJQi15Ei5GzrL641HiyiwS1zES9QmC0U4tEtJS1QNS4M%2BTvS41r49hay2ZUwjm0/DwEI/nXzXywAOokCVgdDqr1xm3pZWwERjluQa012WMfSfiKQjwIpeAkCc5jujyCN4DCOYCiOl7iOSORMuBRMAhFbDt4JiZOA8QWglyV0TyGqWN1sQZCzVih32HaABCSgNRKlf2QMCbUO5ZTIztzsLtTsSOIWUS0pAuzRov1Rca1uzv1u5tKJnCSixYgMNoBnmhliZ5nlSIMpMEpbbpSLOLnXMCnuWkRNwPwQuPqiXFdSH0OV1gn203G24bTOX0OuBh4MNROF4D%2BBnD4fkfCxAIRLJbkzOuIn/P06xJRuBvnLBuZTcWcfEAetfO/DwjAwWt2kvNgCS7EDqgSBRu5MYS3gPzou5FbgZP8DKaECzR8cqc6xsusJkuj7phiZqAMCjheQyZvMHPTK8gAyw3BFnDQhawfzxAkWwhgUeZ60G14IACyWYo86ml%2BZ5ThZ1ry%2BQ5wfCtns8kGo80I0YQilZvnI0YmZtbr7znzXrUn64VrIE%2BbhbtLxbGnB4imNh%2B9tsajuiHDSbWnjUm%2BAbKXDctAga5XZ5YULnGB2i0YHnHwXAhAGXuRYmh%2B5Dl%2B/CC8/AJoo8SXljuhBAHYVXx2Y5bJ9hVXZ248YmuQgS%2Bme13qUGYn26fqmYNnM3254UdKXXIn3HHEdKTF/B2b/WQLgRKQobr3cLJ0JXcI1uhTHEVbVcvLI6tYp1519gVwizISDnEn%2BXeElyPNKer0LGB44PF1F4V4W3a4UDAQT7r474GT9NBWTNKN2IrNGNumWNlHrdrT8ryrdPAYQYV9FNOdUHp9FHY9d9xEwMSpC9ndRztHuiLGd9Ab0sDIx9cLQDf0/jvPxE4M/jZhoP14D7WWePxEZE0d/4ABu9qvqjUvuWQLKB6zP1j4ivuW/j2vb0Q7X0t7ZHRR973bj7GvuWxE3jKvTvav%2BPtDyhlzTDeP5v9GaZqEkudtEvIAr7%2Bd77Q9BAn7jAUrVktKVpcIec80nj33f7usJwGorDi4pbTLkuNUY8IP6Yz21NrH%2BMGu48unT3BxZlYmOqJAuw0YWYod%2Bno8i66NLne7s7R%2B0lA3TAnnBlQ0glCyWv93NEWvfUb4J33FQPNEYmAoGzjDZYXoHYCh24Z1mA1uZwlYsUimGmgaCQCH0U2f1gmYdwo/13lvE/W4y5N/qfXx8clbgF1bFnNr%2Bz6SwQp3yTTXVkitWvK9KhhvRnB5AtxLBMuWT60JKka7PHAzQ8yk9Is5POgHqHvKM1/Q0NJMNhQ0B2VLAjlZyllVIDuU4qfhPyq5UCpIASQNNCKrNCUZX0DAPUPkAECC5PAAA0tgDhBOA1AYoGIHyDhCSgAAmtoFfB8gFiAiYACYDQCugz8dSawMACtDABMAYoCcPYGcBFVJcxIM6gKB%2BDvQCAOqM/DFQMAmAoq8seABAFhBhViAygiwCYOsHyBRABg/0B5TtDeVhApAUAPoBMBfhOa1ICcK5XIAaAguoNBgJIH2p7BjAAQpfowEIF5VgA1gzwYFQsHe1rBUQWwROFSGOC2AoRCJvFQ8FeCNAPg94H4IiFfggh8QUIdsAOr6B5A%2BQqIS5QiGxD4hdlDYF%2BGYGv5BGqAUwA2AMA1CzAFgBIIpn0H0BrBBweyvgNID1C3KBgSUOqBpBAhehZAgKhsErAgA20eA/QIoBMAsAFABBCYYQPcqeD3AFgRYbgLgCwAYAiAYKtIPoBkAKAq0a4ZMGAANUMhDzX8JQBSCECUg8QcND0JMBfCkMAoFIF0K8C/DSAUgtgBIAFAMAOuhArAOIJZg1hqhJgLADcBsByB/B%2BAdUC0C1iIiIhXAZoD8HCFeCuaDlfwZ%2BF3zhoggWAQgXBG2FIjSAKWV%2BFoAWKojgAn4fKuQNEBWBQwAANUY6KY6hoIsQIlVLApVlAwo9KjoD0D%2BCcqlgOQQVTUHOBLAeAFIDWEgAbBLmiIktOOBKiqCVkTgBwNqkYH0lRCblFLKSCwBqjqog1L4mVUCBjV5g1VdAJtUmD2CuqlQIoA6N6AtUPRLo%2BqjaKZAjUugXoiwFNVaAbU5qKwV0YsBWrjV7By1WajUDqoKANgVAlKmMP0BOU9h/gg4WcE4CNVkIPUZCIoCfTIBFIDVfgOLggAzCLahweQGsBMAnCKBVwjCDcJoFSDWxjw54Q4NeEpR3hnw74QGlBH/Dw0gI4EWQN6HgjYkUImERiMwDwjM6oIlEXIPRERDMR7UHEYQPxGS5fghA/bqSIaEqizoAaKkUSNIC0jQRjIvIJgBZEriyQWVDYFQG5EaA%2BRmAAUT0yFGqAkqcgVKhKM0BSjCB5gewbEIVH6jlRqo%2BABqOX6tIGBPUHUdfD1GuBDRZwHqAKBwFEDzRbYUweqIDFgxzB5VEMfYNiCRjkxboioG1UIk%2BjGwfo0MSVWmojB2q8YuieGKWAkSFq8YmaoxJjGJj5qW1VMbMJTHrCsxkwogQYDzEFiixJY4AGWNAEHBKx4UGsXMIOD1jGxD45YasMMDrDNhpAbYWsJEkHCTANYewU2NIBnDLh4qXQe2LOqdjiAiQJwfoHzGFjixpY8sXJIcDGCvxoohQOKNUD/jMqMojqueNmFtgLAimM6DwF%2BEZjhJ%2BwgwNoIIC6CRE4kpyVJJkkViqxQQB4VeSBBfpVJnIoKhZJ%2BBWTMpdkrIQ5IknOTpJrkysR5JFHJVvJaVPydKIiFASgpNNEKfYLCm3BIpQkggTmNik6CeINTRyZJJcmyT5JEADKTZOsgqTzxak0gLFm0STBrRpI7SbpM0n6SjBRwzSU2PUlrDSRBwXqREIMmzTyBGY9ydmKOmbSdpDI38LLEUBAA) contains all the other uses and corner cases (plus the kitchen sink). _Really this one probably isn't worth spending a lot of time on_. The two items of note in it are **a)** an extension of the `PostOffice` into the `Service` object (allowing for multiple registrations to a single topic string) and **b)** the `LongRunningCallback` and its corresponding test code (showing that callbacks block object destruction and touching on the member variable ordering requirement mentioned above).

----

I will admit - `Connection.h` is not _ten_ or _twenty_ lines of code.
```
$ cloc Connection.h 
       1 text file.
       1 unique file.                              
       0 files ignored.

github.com/AlDanial/cloc v 1.82  T=0.01 s (180.1 files/s, 24856.7 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
C/C++ Header                     1             21             17            100
-------------------------------------------------------------------------------
```
It's _100_ lines of code. And it's code that is going to touch _every_ C++ use of uTransport.

I've already said I lean towards a mechanism like this as, in my opinion, the risks of misusing it are smaller than the likelihood of a developer forgetting to explicitly call a cleanup / unregister function. That is, after all, why standard library classes like `std::unique_ptr` and `std::lock_guard` exist. But I'm also open to being convinced otherwise.

Reply:
Regarding the nature of UTransport having methods versus UTransport being a namespace, and having member functions that take a session object (like syscalls and file descriptors), in C++, namespaces can be added to after their primary declaration. Class declarations cannot be added to. They are inherently write-once. Using a namespace and a member session object is future-proofing in design. This is unlike python, for example, where objects can be modified. I'm not sure about Java on this matter.

Reply:
Regarding registerListener, I am concerned by the fact that the RPC 'listener' has a very different role than a subscriber listener. I don't think the two should be bundled into the same registration function. For one thing, RPC server worker functions necessarily have a response (or the general expectation to provide a response). This means there will be a callback on the same transport handle as the incoming request. This response handle can be provided through thread-local storage, but it is much more straightforward for the response handle to be passed into the RPC server worker function. I makes sense for the RPC server worker function to have a different argument signature than a subscription. Another reason that RPC server worker functions are likely to be different from subscriber workers is their thread pooling context. It seems likely that RPCs should be processed by a thread pool. It seems possible that a subscription server's work should be done in a dedicated thread, or a dedicated cluster of threads.

Reply:
Self-terminating connection handles are returned now when registering callbacks.

Reply:
> Class declarations cannot be added to.

I don't know what is meant here.

> I am concerned by the fact that the RPC 'listener' has a very different role than a subscriber listener.

These modes can be determined from the URI used to register, should the transport implementation need to handle them differently. Not all transports will distinguish between the two modes, nor does uProtocol require that they do.